In [2]:
import numpy as np
import matplotlib.pylab as plt

In [3]:
def softmax(a):
    c = np.max(a)
    exp_a = np.exp(a-c)   #개선
    sum_exp_a = sum(np.exp(a-c))
    y = exp_a/sum_exp_a
    return y

In [4]:
def cross_entropy_error(y, t):
    if y.ndim == 1:
        t = t.reshape(1, t.size)
        y = y.reshape(1, y.size)
        
    # 훈련 데이터가 ONE-HOT 벡터라면 정답 레이블의 인덱스로 반환
    if t.size == y.size:
        t = t.argmax(axis=1)
             
    batch_size = y.shape[0]
    return -np.sum(np.log(y[np.arange(batch_size), t] + 1e-7)) / batch_size

In [5]:
class simpleNet:
    def __init__(self):
        np.random.seed(5) #항상 동일한 랜덤값 출력을 위함 (수업 목적)
        self.W = np.random.randn(2,3) # 정규분포로 초기화

    def predict(self, x):
        return np.dot(x, self.W)

    def loss(self, x, t):
        z = self.predict(x)
        y = softmax(z)
        loss = cross_entropy_error(y, t)

        return loss

In [6]:
net = simpleNet()

In [7]:
x = np.array([0.6, 0.9]) # 입력
t = np.array([0, 0, 1]) #정답, TARGET

In [8]:
net.W #어떤값으로 초기화 되어있는지 확인

array([[ 0.44122749, -0.33087015,  2.43077119],
       [-0.25209213,  0.10960984,  1.58248112]])

### 입력값을 이용하여 FORWARD 방향 연산해보기

In [9]:
pred = net.predict(x)

In [10]:
print(pred)

[ 0.03785358 -0.09987323  2.88269572]


In [11]:
y = softmax(pred)

In [12]:
print(y)

[0.05243789 0.04569106 0.90187105]


In [13]:
np.argmax(y)

2

### 동일한 입력값과 정답을 이용하여 loss 구해보기 

In [14]:
net.loss(x, t)

0.10328361502771058

### 생각해보기

* 일단 세번째 (인덱스 #2) 값이 제일 크니 맞춘 것은 같다. 😀

         2.88269572

In [15]:
print(pred)

[ 0.03785358 -0.09987323  2.88269572]


* Softmax 를 이용하여 확률로 해석을 해봄

         0.90187105 --> 세번째(인덱스 #2) 값이 90% 확률로 정답일 것이다.

In [16]:
y = softmax(pred)
print(y)

[0.05243789 0.04569106 0.90187105]


* Argmax 함수로 최종 분류 확인

        np.argmax(y)의 값이 2인 것을 확인

In [17]:
np.argmax(y)

2